Zend/zend_object_handlers.c

```diff
@@ -604,6 +617,73 @@ ZEND_API uint32_t *zend_get_recursion_guard(zend_object *zobj)
 	return &Z_GUARD_P(zv);
 }
 
+ZEND_COLD static void zend_typed_property_uninitialized_access(const zend_property_info *prop_info, zend_string *name)
+{
+	zend_throw_error(NULL, "Typed property %s::$%s must not be accessed before initialization",
+		ZSTR_VAL(prop_info->ce->name),
+		ZSTR_VAL(name));
+}
+
+static ZEND_FUNCTION(zend_parent_hook_get_trampoline);
+static ZEND_FUNCTION(zend_parent_hook_set_trampoline);
+
+static bool is_in_hook(const zend_property_info *prop_info)
```

Comment:
This is an interesting idea. It won't solve the problem fully, in that it would be possible to access sibling properties if the flag is not bound to specific properties. Unfortunately, this also seems like a common scenario (prop->a::get accesses prop->b, prop->b::get accesses prop->a).

Maybe there are other ways to improve this.

Reply:
Can you just use the flag _and_ allocate an extra temporary for hook functions in the first slot, containing a pointer to the property_info?
So, when calling hook functions, you set `Z_PTR_P((zval*)(execute_data + 1))` to the property_info in addition to the flag.
Doesn't need enumerating all the special cases here then.

Reply:
Essentially, what we need is a link from the `zend_function` to the corresponding `zend_property_info`. Unfortunately, using a `T` doesn't work great with the `zend_call_function()` API when calling the hook from the object handler. Instead, I will try to link the property from `zend_function` directly.

Reply:
Please check https://github.com/php/php-src/pull/13455/commits/71b1f604031d8a7f2d984de69e0a7ed59693568d. Does this look like a better solution to you?

Reply:
Since this is only used when hooks are used, and the fact that the hook function pointer array is stored separately, I think it's best to combine the two kinda like Bob suggested. Except I suggest it with less pointer magic and just have a structure like this (conceptually):

```c
struct zend_function {
// ...
    hook_data *hook_data;
};

struct hook_data {
    const zend_property_info *prop_info; /* The corresponding prop_info if this is a hook. */
    zend_function *hooks[2];
}
```

Reply:
@nielsdos The problem is that I need to fetch the property info from the function on the stack, i.e. `EG(current_execute_data)`, but from there I don't have access to the `zend_property_info.hooks`. If I did, then the problem wouldn't exist in the first place. :smile:

Reply:
No worries. php-src sometimes feels specially crafted to confuse me. :sweat_smile:

Reply:
I forgot about reflection. It required some more hacks... https://github.com/php/php-src/pull/13455/commits/a8d7d7cc057851f983a77aa3f18989f5b6a65047

Reply:
We could avoid the hack by re-introducing a special case in is_in_hook(), but implementing it in reflection:

```
bool is_in_hook(const zend_property_info *prop_info) {
    if (UNEXPECTED(!EX(func)->common.prop_info)) {
        return zend_reflection_is_in_hook(prop_info);
    }
    ...
}

// php_reflection.c
bool zend_reflection_is_in_hook(const zend_property_info *prop_info) {
    // check that EG(func) is setRawValue
    // fetch prop_info from the reflection object
}
```

I'm not sure that's better. The change otherwise looks good to me. The hack is okay if that's the only barrier to this solution.

Reply:
Can we have a ZEND_API `zend_property_set_hook(prop_info, execute_data, func_backup)` and `zend_property_release_hook(prop_info, execute_data, func_backup)` or something? Rather than doing the very hacky thing in reflection?

I do like the current impl of is_in_hook though.

Reply:
I can move the logic into a function, sure. I'll also link prototype to `self` when there is no parent, so that we can skip the ternary in `is_in_hook()`.